PR to resolve Ansible 2.0 issues

DavidjohnBlodgett opened the pull request:
@benbp Just pushing this up to expose what I did and get feedback... this passed a golden run where I faked the expected files for the oem(like you suggested). However, I would like to track down where we get the below files:

ami
flashupdt
socflash
syscfg

because they do not seem to come into the expected directory from the .deb referenced in the README. (In addition, we never instruct the user on extracting that file... so I would like to add the instructions for this as well.)

I plan to clean up the README instructions to reflect my changes and hopefully instructions for the missing pieces I hit.

DavidjohnBlodgett committed:
Implemented wrapper shell scripts to replace Ansible playbooks in order to support Ansible 2.0

squashed commits

added set and comments, rm old yml files

added example.sh

updated readme

more updates to readme

some more updates to readme

readme and wrapper updates and travis bumped to 2.0 install

fixing travis

fixing path issue

cleanup after rebase

travis edit

added version to travis ansible install

use pip for ansible

DavidjohnBlodgett commented:
In addition, I apologize for what looks like a lot of changes... I did a lot of experimentation on this branch and those changes were slurped into this, but there is really only two new files that are the major changes.

stuart-stanley commented:
@DavidjohnBlodgett:
You are correct on those files. Alas, I haven't been able to find the breadcrumb trail about where they came from. Presuming I could ever find them, I had been thinking it would be nice to setup conditionals for building particular overlays if the files ARE there. Something along the lines of "'oh, I see ami and socflash in <some-dir>, so I can build the quanta overlay. Yay!".

There is also the storcli deb package (also limited by license) that needs to get put into the discovery.overlays.cpio.gz file if it is there. I might be wrong, but I believe it can't sit in an overlay like the quanta and intel flashing tools, since it is used at, well, discovery time :).

stuart-stanley commented on roles/ipxe/tasks/main.yml:
```diff
@@ -58,12 +61,12 @@
     chdir: "{{ ansible_env.HOME }}/ipxe/src"
```
You might want to move this to go in a subdir of the build-space instead of the users $HOME. Everything else goes to /tmp/on-imagebuilder/its-working-dir, but this one plops itself down elsewhere. There is already an ipxe_build_artifact_path defined in hosts, so adding an ipxe_build_working_path to another directory within the build-space is pretty simple. Aside from just being odd having one part of the work done in a different place than the others, it really bites on a mounted home directory :)

DavidjohnBlodgett replied:
@benbp do you know if we did these ipxe builds in the users home dir due to some permissions thing? or any other issue that may not be obvious to me? If not would you like me to generate a new story to implement the suggestion above? @stuart-stanley thanks for the feed back!

benbp replied:
Yeah it probably makes sense to add a new story for this.

stuart-stanley commented on build_all.sh:
```diff
@@ -0,0 +1,40 @@
+}
+
+# initrd
+arg1=common/initrd_wrapper.yml;
```
nit: arg# does not tell me much (and I can see when I look at the PlayBookRunner line). "playbook" instead of arg1, "config" instead of arg2, and "provisioner" instead of arg3?

benbp commented:
I agree with @stuart-stanley's comments. I get that the script is trying to add a convenient layer of abstraction, but it makes it way harder to understand what is actually happening at a glance. I'm inclined to say it's easier to just express things directly like:

```
# build_discovery_images.sh
sudo ansible-playbook -i hosts common/initrd_wrapper.yml -e "config_file=vars/initrd.yml provisioner=roles/initrd/provision_initrd"
sudo ansible-playbook -i hosts common/basefs_wrapper.yml -e "config_file=vars/basefs.yml provisioner=roles/basefs/provision_rootfs"
sudo ansible-playbook -i hosts common/overlay_wrapper.yml -e "config_file=vars/overlay.yml provisioner=roles/overlay/provision_discovery_overlay"
```

benbp commented:
Also do these dependency updates fix currently broken builds? If so, can you submit them as a separate PR so we can merge that part a little quicker?

benbp commented:
@DavidjohnBlodgett I actually have copies of all those files I can send to you, but are simple mock files not working for you?

DavidjohnBlodgett commented:
@benbp yes, these dependency bumps fix what seems to be issues with the current rev's available  on apt. I'll post a separate PR with the changes I made now.

DavidjohnBlodgett commented:
@benbp here is what you asked for https://github.com/RackHD/on-imagebuilder/pull/24

DavidjohnBlodgett commented:
@benbp @stuart-stanley understood guys, I'll simplify it to just direct execution of the commands. I only wrote it up that way cause I figured I'd get set on fire for not using some loop logic, etc... lol

benbp commented on build_all.sh:
```diff
@@ -0,0 +1,11 @@
+
+# this script replaces the all.yml playbook due to a variable scoping issue
+# that appeared with Ansible 2.0.0.2
+SCRIPT_DIR=$(cd $(dirname $0) && pwd)
```
What is this for? Also, I don't think we really need this comment. The commit message/PR explains it for anyone who is really curious.

DavidjohnBlodgett replied:
guilty of leaving code in that I was thinking maybe useful later to create an absolute path... I'll purge it np

benbp commented on build_all.sh:
```diff
@@ -0,0 +1,11 @@
+# that appeared with Ansible 2.0.0.2
+SCRIPT_DIR=$(cd $(dirname $0) && pwd)
+
+# build_discovery_images.sh
```
Make this comment say something useful?

benbp commented on build_all.sh:
```diff
@@ -0,0 +1,11 @@
+sudo ansible-playbook -i hosts common/initrd_wrapper.yml -e "config_file=vars/initrd.yml provisioner=roles/initrd/provision_initrd"
+sudo ansible-playbook -i hosts common/basefs_wrapper.yml -e "config_file=vars/basefs.yml provisioner=roles/basefs/provision_rootfs"
+sudo ansible-playbook -i hosts common/overlay_wrapper.yml -e "config_file=vars/overlay.yml provisioner=roles/overlay/provision_discovery_overlay"
+sudo ansible-playbook -i hosts common/ipxe_builder.yml -e "config_file=vars/ipxe.yml"
```
Maybe add a comment above differentiating this from the discovery image builds?

DavidjohnBlodgett replied:
should I just label it "build static files" ?

benbp replied:
How about "build default discovery images" ?

DavidjohnBlodgett commented:
once I get the dependency update PR merged, I'll rebase this and trouble shoot why travis hates me...

benbp commented:
A couple things:
- Can you update the README too since the way of doing this has changed?
- Will you add `set -e` to the scripts so that if one ansible playbook run fails it doesn't try to do the other one?
- Also, you can just delete oem.yml and all.yml now, right?

DavidjohnBlodgett commented:
@benbp understood I'll make these changes, and yes~ I planned to remove the oem.yml and all.yml once I was finally happy. kept them just for reference.

DavidjohnBlodgett commented:
@benbp I updated the readme just now, and could use a second set of eyes on it to make sure I properly corrected for the changes~ in addition, I believe I have fixed the other issues you voiced. Just need to sort travis out

benbp commented on README.md:
````diff
@@ -69,14 +69,14 @@ Getting started
 ```
 
-**All ansible runs must be done on the host machine that is building the images.** This is because we use
+**All build runs must be done on the host machine that is building the images.** This is because we use
````
"All builds must be done" ?

DavidjohnBlodgett replied:
ugh, getting sloppy at the end of the day... sorry, I'm all over it

benbp commented on README.md:
```diff
@@ -93,21 +93,16 @@ If you have experience with Ansible, some of these steps will be familiar:
-      vars:
-        - config_file: vars/my_overlay.yml
-        - provisioner: <path to role directory, e.g. roles/overlay/provision_discovery_overlay>
+    sudo ansible-playbook 
```
Add some line breaks? e.g. 

```
sudo ansible-playbook \
         -i hosts ... \ 
         etc.
```

benbp commented on build_all.sh:
```diff
@@ -0,0 +1,14 @@
+set -e
+
+# build initrd
+sudo ansible-playbook -i hosts common/initrd_wrapper.yml -e "config_file=vars/initrd.yml provisioner=roles/initrd/provision_initrd"
```
Do you think it would be clearer to line break here as you have done in the README example?

benbp replied:
I don't really know.

benbp commented on common/basefs_wrapper.yml:
```diff
@@ -13,13 +13,14 @@
     - ../{{ config_file }}
   roles:
     - ../{{ provisioner }}
+    - ../roles/basefs/provision_rootfs
```
This line should not exist.

benbp commented on common/overlay_wrapper.yml:
```diff
@@ -11,13 +11,14 @@
     - ../{{ config_file }}
   roles:
     - ../{{ provisioner }}
+    #- ../roles/overlay/provision_discovery_overlay
```
Delete this line.

benbp commented:
Two more things:
- Add an Ansible > 2.0 in the Requirements section of the README.
- Update the travis.yml file to install 2.x instead of 1.9 (https://github.com/RackHD/on-imagebuilder/blob/master/.travis.yml#L6)

tldavies commented:
Need to git squash fixing commits.

DavidjohnBlodgett commented:
@tldavies understood, will work on that after I shake out what's breaking travis

DavidjohnBlodgett commented:
@benbp result!

benbp commented:
Still needs a squash :)

Do the newly built images work for a simple discovery trial run?

iceiilin commented:
@DavidjohnBlodgett Is there any consideration about changing wrapper yml to scripts?

DavidjohnBlodgett commented:
@iceiilin hey Lyne, I had brought up moving other parts of this process to shell scripts with @benbp earlier but it was decided that we should just focus on the above fix. There is no other plans to change this unless another issue arises with Ansible, I hope that answered your question. :)

DavidjohnBlodgett commented:
@tldavies @benbp commits sorted.

DavidjohnBlodgett commented:
@benbp to answer your earlier question, yes I have just verified discovery was not broken with this change.

benbp commented:
@DavidjohnBlodgett cool. I know you just squashed as a second commit, but I was hoping you would squash everything down into a single commit that just explains the entirety of what you've done here, rather than having 2 commits here, the first of which is mostly just WIP code that gets deleted.

DavidjohnBlodgett commented:
@benbp oh sorry, understood. I'll circle back and fix that.

AlliumApotheosis commented on README.md:
````diff
@@ -69,14 +69,14 @@ Getting started
 ```
 
-**All ansible runs must be done on the host machine that is building the images.** This is because we use
+**All builds runs must be done on the host machine that is building the images.** This is because we use
````
All build runs must be executed locally on the machine that is building the images. ?

DavidjohnBlodgett replied:
it did not seem appropriate to use "Ansible" on this line any more, because we don't run an ansible playbook directly anymore. Instead, we execute a shell script that then invokes ansible with additional arguments.

rolandpoulter commented:
:+1:

tldavies commented:
:+1:

DavidjohnBlodgett commented:
@RackHD/corecommitters

DavidjohnBlodgett commented:
I think this is merge ready~

benbp commented:
:+1:
